crates/bevy_pbr/src/light.rs

```diff
@@ -204,6 +203,107 @@ pub enum SimulationLightSystems {
 // The z-slicing method mentioned in the aortiz article is originally from Tiago Sousa’s Siggraph 2016 talk about Doom 2016:
 // http://advances.realtimerendering.com/s2016/Siggraph2016_idTech6.pdf
 
+#[derive(Debug, Copy, Clone)]
+pub enum ClusterFarZMode {
+    // use the camera far-plane to determine the Z-depth of the furthest cluster layer
+    CameraFarPlane,
+    // calculate the required maximum Z-depth based on currently visible lights.
+    // Culls lights from fragments better, speeding up GPU lighting operations
+    // at the expense of some CPU time as the cluster index list is better filled.
+    MaxLightRange,
+    // constant max Z-depth
+    Constant(f32),
+}
+
+#[derive(Debug, Copy, Clone)]
+pub struct ClusterZConfig {
+    // far depth of the nearest cluster layer
+    pub first_slice_depth: f32,
+    pub far_z_mode: ClusterFarZMode,
+}
+
+impl Default for ClusterZConfig {
+    fn default() -> Self {
+        Self {
+            first_slice_depth: 5.0,
+            far_z_mode: ClusterFarZMode::MaxLightRange,
```

Comment:
This is good for when we have storage buffers, but how does this behave for WebGL2 when we have to use uniform buffers and only have 16384 indices available?

Reply:
looking specifically at choosing MaxLightRange vs CameraFarPlane (which is equivalent to the current setup), there isn't a one-size-fits-all best answer but i think MaxLightRange is better.

the CPU time for pushing 16k indices is of order 1ms when all the outstanding optimisations are merged, so i'll focus on render time.

- for a scene with a single light / a few lights, render time will be unchanged / fractionally better, but light calculations will very quick regardless.

- for a scene with a lot of lights with a broad distribution of depths, so that most lights only occupy a small number of z-slices, render time will be improved as we will reduce the average light count per cluster by fitting the slices better to the scene

- for a scene with a lot of lights with a narrow distribution of depths, render time will be worse as MaxLightRange will make the cluster tilesize expand in x/y, so average light count per cluster will increase.

There is an argument to be made for using Single cluster mode, effectively making all the clustered forward rendering opt-in, at which point users will be forced to think about and select the right mode for their setup. But it clearly gives a better first impression if scenes with a lot of lights are fast out of the box. Then, I think a broad range of depths is a more common setup than 100s of lights at a similar depth.
For the 100s of lights at a similar depth case, the better solution is to use fewer z-slices rather than to change the z-mode. I can only think of very contrived cases where a constant far plane is better in any way, but i didn't want to stop supporting it without being sure.

Reply:
That was stated clearly, thanks. I agree with the evaluation and as the many lights at the same depth case is likely more rare, it can be considered then. But perhaps we can add some documentation suggesting use of 1 or few depth slices for that case, as top-down / isometric 3D games are not _too_ unusual. I'll make another documentation suggestion.

Reply:
Just to note that this thread of discussion is resolved but there is no resolve button for me to push.